sql/opt: add implicit SELECT FOR UPDATE support for UPSERT statements

nvb opened the pull request:
Fixes #50180.

This commit adds support for implicit SELECT FOR UPDATE support for UPSERT statements with a VALUES clause. This should improve throughput and latency for contended UPSERT statements in much the same way that 435fa43 did for UPDATE statements. However, this only has an effect on UPSERT statements into tables with multiple indexes because UPSERT statements into single-index tables hit a fast-path where they perform a blind-write without doing an initial row scan.

Conceptually, if we picture an UPSERT statement as the composition of a SELECT statement and an INSERT statement (with loosened semantics around existing rows) then this change performs the following transformation:
```
UPSERT t = SELECT FROM t + INSERT INTO t
=>
UPSERT t = SELECT FROM t FOR UPDATE + INSERT INTO t
```

I plan to test this out on a contended `indexes` workload at some point in the future.

Release note (sql change): UPSERT statements now acquire locks using the FOR UPDATE locking mode during their initial row scan, which improves performance for contended workloads when UPSERTing into tables with multiple indexes. This behavior is configurable using the enable_implicit_select_for_update session variable and the sql.defaults.implicit_select_for_update.enabled cluster setting.

nvb committed:
sql/opt: add implicit SELECT FOR UPDATE support for UPSERT statements

Fixes #50180.

This commit adds support for implicit SELECT FOR UPDATE support for
UPSERT statements with a VALUES clause. This should improve throughput
and latency for contended UPSERT statements in much the same way that
435fa43 did so for UPDATE statements. However, this only has an effect on
UPSERT statements into tables with multiple indexes because UPSERT
statements into single-index tables hit a fast-path where they perform a
blind-write without doing an initial row scan.

Conceptually, if we picture an UPSERT statement as the composition of a
SELECT statement and an INSERT statement (with loosened semantics around
existing rows) then this change performs the following transformation:
```
UPSERT t = SELECT FROM t + INSERT INTO t
=>
UPSERT t = SELECT FROM t FOR UPDATE + INSERT INTO t
```

I plan to test this out on a contended `indexes` workload at some point
in the future.

Release note (sql change): UPSERT statements now acquire locks using the
FOR UPDATE locking mode during their initial row scan, which improves
performance for contended workloads. This behavior is configurable using
the enable_implicit_select_for_update session variable and the
sql.defaults.implicit_select_for_update.enabled cluster setting.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53132)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53132#-:-MFBjMpXC57Ow3po_wZQ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

nvb committed:
workload/indexes: add --cycle-length flag

Controls the number of keys repeatedly accessed by each
writer through upserts. Mirrors the same flag in `kv`.

nvb committed:
sql/opt/exec: re-enable locking policy EXPLAIN non-verbose attributes

Reverts a portion of #53003. These attributes are not included when they
are not interesting, but when they are included, they are very interesting
and deserve to be surfaced.

nvanbenschoten commented:
> I plan to test this out on a contended indexes workload at some point in the future.

To test this, I ran a quick, unscientific experiment on my laptop. Using the `indexes` workload's new `--cycle-length` flag, I created a setup with a decent amount of contention. `--concurrency` was set to 32 and `--cycle-length` was set to 32. I then varied the number of secondary indexes while switching implicit SELECT FOR UPDATE on and off.

![throughput](https://user-images.githubusercontent.com/5438456/90807196-1aeb2a00-e2ec-11ea-92a4-79029273d1f0.png)

![latency](https://user-images.githubusercontent.com/5438456/90807201-1c1c5700-e2ec-11ea-8bef-4845e59f377b.png)

For `--secondary-indexes=0`, implicit SFU made no difference because of the fast-path discussed above. For all other cases, we see the same [trends that we saw with implicit SFU for UPDATE statements](https://www.cockroachlabs.com/blog/when-and-why-to-use-select-for-update-in-cockroachdb/). First, we see that throughput is a little over 50% higher using implicit SFU. Second, we see that tail latency is about 90% lower using implicit SFU.

Again, this was an unscientific experiment, but it makes a pretty strong claim that this change is beneficial for contended UPSERT statements.

nvanbenschoten commented:
TFTR!

bors r+
